### PR TITLE
Change import/export format from CSV to JSON

### LINKED ISSUE #4: Change data import/export format to JSON
### Context

We're currently exporting data to CSV but it's very ugly for nested data structures like the list of `wateringTimes` (and later the list of plant actions). We work around that by encoding the list of watering times as a semicolon-delimited list of ISO8601 date strings, which the CSV parsing function needs to be aware of. This is complex and adding more data to the plant structure will make it more complex.

We should also include the migration index - which isn't currently included in the CSV output. If a user exports their data at migration index 2, and then a few app updates are released which increment the migration index to 4, and then the user imports their old data export (at index 2), their database will be populated with data in the form expected at migration index 2 which may break the running code (which expects it to be at migration index 4).

We can solve that problem easily if we run the extra 2 migrations, but since our migration index is already at 4 it looks like we don't need to, so the app won't.

### Proposed Solution

1. Export data as JSON including:
    * all plant objects in their current form
    * the current ID counter (so we don't generate IDs that conflict with existing ones)
    * the current migration index
2. Modify the import function:
    * if the migration index in the JSON is not the current latest, reject it (users can contact me to convert their data if they need it)
    * to parse the data as JSON

### PR DESCRIPTION
**Pull request by wmadden**
Closes #4

### CODE REVIEW

File: clients/web-app/src/data/exportData.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  //   plant.id,
+  //   plant.name,
+  //   plant.wateringPeriodInDays,
+  //   `"${plant.wateringTimes.map((date) => date.toISOString()).join(";")}"`,
+  //   plant.timeOfDeath ? plant.timeOfDeath.toISOString() : "",
+  // ]);
+  // const content = [headerRow, ...bodyRows].join("\n");
+
+  // return new Blob([content], { type: "text/csv;charset=utf-8" });
+
```
Comment (reydelo) [inline clients/web-app/src/data/exportData.ts:27]: no need to keep the code from generateCSV here

File: clients/web-app/src/data/importData.test.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      importData(jsonString);
+    }).toThrowError(/not valid JSON/);
+  });
+
+  it("throws an error if the encoded value is not an object", () => {
+    const jsonString = "null";
+
+    expect(() => {
+      importData(jsonString);
+    }).toThrowError(/does not contain a valid export object/);
```
Comment (reydelo) [inline clients/web-app/src/data/importData.test.ts:47]: what is this syntax? I assume like regex to mean `/contains this string/`? I can't find the documentation for `toThrowError()` in jest

Comment (reydelo) [reply clients/web-app/src/data/importData.test.ts:47]: [Found it](https://jestjs.io/docs/en/expect#tothrowerror), it's an alias for `toThrowError`

_Note: Being verbose and adding notes here as a learning strategy for me_
```
test('throws on octopus', () => {
  function drinkOctopus() {
    drinkFlavor('octopus');
  }

  // Test that the error message says "yuck" somewhere: these are equivalent
  expect(drinkOctopus).toThrowError(/yuck/);
  expect(drinkOctopus).toThrowError('yuck');

  // Test the exact error message
  expect(drinkOctopus).toThrowError(/^yuck, octopus flavor$/);
  expect(drinkOctopus).toThrowError(new Error('yuck, octopus flavor'));

  // Test that we get a DisgustingFlavorError
  expect(drinkOctopus).toThrowError(DisgustingFlavorError);
});
```

File: clients/web-app/src/utilities/lang/Override.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+export type Override<
+  Type1 extends { [key: string]: any },
+  Type2 extends { [key: string]: any }
+> = Omit<Type1, keyof Type2> & Type2;
```
Comment (reydelo) [inline clients/web-app/src/utilities/lang/Override.ts:4]: I don't understand what this does, would you help me understand?

File: clients/web-app/src/data/exportData.ts (hunk 1 above)
Comment (wmadden) [reply clients/web-app/src/data/exportData.ts:27]: Oh nooooo

File: clients/web-app/src/data/importData.test.ts (hunk 2 above)
Comment (wmadden) [reply clients/web-app/src/data/importData.test.ts:47]: FYI you said ‘toThrowError()’ is an alias for itself. You can find documentation for all matchers in the jest documentation

File: clients/web-app/src/utilities/lang/Override.ts (hunk 3 above)
Comment (wmadden) [reply clients/web-app/src/utilities/lang/Override.ts:4]: If you combine two types that include the same key but different values the compiler complains. This returns a type which combines the two given types, where keys on the right override keys on the left

**Pull request merged**